// @overvis/opcb-api/dist/model/modbus-free-ranges.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.remove = exports.update = exports.create = exports.list = exports.ITEM_SCHEMA = void 0;
const opcb_ts_shared_1 = require("@overvis/opcb-ts-shared");
const conf_crud_1 = require("./conf-crud");
exports.ITEM_SCHEMA = opcb_ts_shared_1.MODBUS_FREE_RANGE_ITEM_SCHEMA;
function toListItem(rec) {
    return {
        device: rec.device === undefined ? undefined : parseInt(rec.device),
        function: parseInt(rec.function),
        range: rec.registersRange,
    };
}
function fromCreateItem(item) {
    return {
        device: item.device ? item.device.toString() : null,
        function: item.function.toString(),
        registersRange: item.range,
    };
}
_a = (0, conf_crud_1.createConfCrud)("auth:freeAccessRequests", ["device", "function", "registersRange"], toListItem, fromCreateItem, fromCreateItem, ["device"]), exports.list = _a.list, exports.create = _a.create, exports.update = _a.update, exports.remove = _a.remove;
