// @overvis/opcb-config-manager/dist/start.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.start = void 0;
const tslib_1 = require("tslib");
const pino_1 = tslib_1.__importDefault(require("pino"));
const opcbConfigManager = tslib_1.__importStar(require("."));
let logger;
async function start() {
    logger = (0, pino_1.default)({
        level: "info",
        transport: {
            target: "pino-pretty",
        },
    });
    logger.info("Start config manager in test mode.");
    let configFilePath = "../opcb-deploy/dev-config.json";
    if (process.argv[3]) {
        // set path to test config
        configFilePath = process.argv[3];
    }
    await opcbConfigManager.run(logger, {
        redisConnectString: "redis://127.0.0.1:6379/0",
        manufacturerFile: { path: "../opcb-deploy/DEV_MANUFACTURER" },
        configFile: { path: configFilePath },
        factoryConfigFile: { path: "../opcb-deploy/dev-config-factory.json" },
        sqliteDbPath: "./dev.sqlite",
        sqliteLibDir: "./sqlite",
        dbMigrationsDir: "./db-migrations",
    });
}
exports.start = start;
start().catch((e) => {
    if (logger) {
        logger.fatal(e);
    }
    else {
        // eslint-disable-next-line no-console
        console.error(e);
    }
    setTimeout(() => process.exit(1), 1000);
});
