// @overvis/ts-shared/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimeoutError = exports.timeout = exports.PromiseMonitor = exports.nullToUndefined = exports.notNull = exports.ensureVariant = exports.MAPPERS = void 0;
var mappers_1 = require("./mappers");
Object.defineProperty(exports, "MAPPERS", { enumerable: true, get: function () { return mappers_1.MAPPERS; } });
var types_1 = require("./types");
Object.defineProperty(exports, "ensureVariant", { enumerable: true, get: function () { return types_1.ensureVariant; } });
Object.defineProperty(exports, "notNull", { enumerable: true, get: function () { return types_1.notNull; } });
Object.defineProperty(exports, "nullToUndefined", { enumerable: true, get: function () { return types_1.nullToUndefined; } });
var promises_1 = require("./promises");
Object.defineProperty(exports, "PromiseMonitor", { enumerable: true, get: function () { return promises_1.PromiseMonitor; } });
Object.defineProperty(exports, "timeout", { enumerable: true, get: function () { return promises_1.timeout; } });
Object.defineProperty(exports, "TimeoutError", { enumerable: true, get: function () { return promises_1.TimeoutError; } });
