// @overvis/opcb-virtual-device/dist/tests/utils.js
"use strict";
/* eslint-disable no-console */
/* eslint-disable @typescript-eslint/no-empty-function */
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeWriteMultipleRequest = exports.makeWriteSingleRequest = exports.makeReadRequest = exports.runVirtDevice = void 0;
const tslib_1 = require("tslib");
const opcb_ts_shared_1 = require("@overvis/opcb-ts-shared");
const virtualDevice = tslib_1.__importStar(require(".."));
const server_tools_1 = require("@overvis/server-tools");
async function runVirtDevice(silent = false) {
    let logger = {
        level: "silent",
        fatal: () => { },
        warn: () => { },
        debug: () => { },
        trace: () => { },
        info: () => { },
        error: () => { },
        silent: () => { },
    };
    if (!silent) {
        logger = {
            level: "warn",
            fatal: console.error,
            warn: console.warn,
            debug: () => { },
            trace: () => { },
            info: () => { },
            error: console.error,
            silent: () => { },
        };
    }
    const bus = new opcb_ts_shared_1.RedisBus("redis://127.0.0.1:6379/0", "CFG", logger);
    await bus.clearCmdStream();
    bus.listenCmdStreamForever().catch(console.error);
    await bus.setConfs({
        "virtualDevice:isEnabled": "true",
        "virtualDevice:uid": "111",
    });
    await bus.sendMsg(new opcb_ts_shared_1.redisBus.SyncConfig());
    const [proc, stopVirtDevice, clearDb] = await virtualDevice.run(logger, {
        redisSocket: "redis://127.0.0.1:6379/0",
        sqliteDbPath: "./src/tests/dev.sqlite",
        sqliteLibDir: process.env.SQLITE_LIB_DIR || "./sqlite",
        testMode: true,
    });
    proc.catch(console.error);
    // eslint-disable-next-line @typescript-eslint/no-unsafe-call
    await (0, server_tools_1.timeout)(5000, bus.waitForMsg("done"));
    return [
        bus,
        () => {
            stopVirtDevice();
            bus.abortListen();
            clearDb();
        },
    ];
}
exports.runVirtDevice = runVirtDevice;
function makeReadRequest(modbusId, address, length, func) {
    const buffer = Buffer.alloc(6);
    buffer.writeUInt8(modbusId, 0);
    buffer.writeUInt8(func, 1);
    buffer.writeUInt16BE(address, 2);
    buffer.writeUInt16BE(length, 4);
    return buffer;
}
exports.makeReadRequest = makeReadRequest;
function makeWriteSingleRequest(modbusId, address, value) {
    const buffer = Buffer.alloc(6);
    buffer.writeUInt8(modbusId, 0);
    buffer.writeUInt8(0x06, 1);
    buffer.writeUInt16BE(address, 2);
    buffer.writeUInt16BE(value, 4);
    return buffer;
}
exports.makeWriteSingleRequest = makeWriteSingleRequest;
function makeWriteMultipleRequest(modbusId, address, values) {
    const buffer = Buffer.alloc(7 + values.length * 2);
    buffer.writeUInt8(modbusId, 0);
    // write func
    buffer.writeUInt8(0x10, 1);
    // write address
    buffer.writeUInt16BE(address, 2);
    // write quantity
    buffer.writeUInt16BE(values.length, 4);
    // write byte count
    buffer.writeUInt8(values.length * 2, 6);
    for (let i = 0; i < values.length; i++) {
        buffer.writeUInt16BE(values[i], 7 + i * 2);
    }
    return buffer;
}
exports.makeWriteMultipleRequest = makeWriteMultipleRequest;
