// @overvis/opcb-ts-shared/dist/schema/gsm.schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GSM_SCHEMA = void 0;
var typebox_1 = require("@sinclair/typebox");
exports.GSM_SCHEMA = typebox_1.Type.Object({
    manageGsm: typebox_1.Type.Boolean({
        default: true,
        description: "If enabled, OPCB will manage GSM stick devcie.",
    }),
    apn: typebox_1.Type.String({
        name: "Access Point Name",
        maxLength: 100,
        description: "Is the name of a gateway between a mobile network (GSM, GPRS, 3G, 4G and 5G)",
        default: "internet",
    }),
    apnUserName: typebox_1.Type.String({
        name: "APN User Name",
        maxLength: 64,
        description: "",
        default: "",
    }),
    apnPassword: typebox_1.Type.String({
        name: "APN Password",
        maxLength: 64,
        description: "",
        default: "",
    }),
    operatorPhoneNumber: typebox_1.Type.String({
        name: "Phone number",
        maxLength: 64,
        description: "",
        default: "*99#",
    }),
    simPinCode: typebox_1.Type.String({
        name: "SIM pin code",
        maxLength: 4,
        description: "",
        default: "",
    }),
}, {
    description: "GSM connection configuration.",
    additionalProperties: false,
    default: {
        manageGsm: true,
        apn: "internet",
        apnUserName: "",
        apnPassword: "",
        operatorPhoneNumber: "*99#",
        simPinCode: "",
    },
});
