// @overvis/opcb-ts-shared/dist/schema/lora-sens.schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LORA_SENS_SCHEMA = void 0;
var typebox_1 = require("@sinclair/typebox");
exports.LORA_SENS_SCHEMA = typebox_1.Type.Object({
    enabled: typebox_1.Type.Boolean({
        default: false,
        description: "Enable LoRaSens gateway.",
    }),
    frequencyDeviceKhz: typebox_1.Type.Integer({
        default: 868175,
        description: "Frequency of the device in Khz.",
    }),
    frequencyGatewayKhz: typebox_1.Type.Integer({
        default: 868100,
        description: "Frequency of the gateway in Khz.",
    }),
    bandwidthKhz: typebox_1.Type.String({
        default: "62.5",
        description: "Bandwidth in Khz.",
    }),
    spreadingFactor: typebox_1.Type.Integer({
        default: 7,
        description: "Spreading factor.",
    }),
    codingRateBits: typebox_1.Type.Integer({
        default: 5,
        description: "Coding rate bits.",
    }),
    txPowerLimitDbm: typebox_1.Type.Integer({
        default: 17,
        description: "Tx power limit in dbm.",
    }),
    loraPollingTimeMs: typebox_1.Type.Integer({
        default: 500,
        description: "Polling time in ms.",
    }),
}, {
    description: "LoraSens gateway settings",
    additionalProperties: false,
    default: {
        enabled: false,
        frequencyDeviceKhz: 868175,
        frequencyGatewayKhz: 868100,
        bandwidthKhz: "62.5",
        spreadingFactor: 7,
        codingRateBits: 5,
        txPowerLimitDbm: 17,
        loraPollingTimeMs: 500,
    },
});
