// @overvis/opcb-ts-shared/dist/schema/monitoring.schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MONITORING_SCHEMA = void 0;
var typebox_1 = require("@sinclair/typebox");
exports.MONITORING_SCHEMA = typebox_1.Type.Object({
    enableCommandFiles: typebox_1.Type.Boolean({
        default: true,
        description: "If true, OPCB will automatically mount inserted USB drives and look for command files for config reset and unbind commands.",
    }),
    monitorTemperature: typebox_1.Type.Boolean({
        default: true,
        description: "If true, OPCB will check CPU temperature periodically and report in the UI and API.",
    }),
    monitorDiskSpace: typebox_1.Type.Boolean({
        default: true,
        description: "If true, OPCB will check storage space periodically and report in the UI and API.",
    }),
    monitorVoltage: typebox_1.Type.Boolean({
        default: true,
        description: "If true, OPCB will check power voltage periodically and report in the UI and API.",
    }),
    internetAccessPingIp: typebox_1.Type.String({
        format: "ipv4",
        default: "8.8.8.8",
        description: "IP address to ping to determine if internet connection is active.",
    }),
    checkForExternalIp: typebox_1.Type.Boolean({
        description: "If enabled, OPCB will periodically check for its own external IP address using https://www.ipify.org/ API.",
        default: true,
    }),
}, {
    additionalProperties: false,
    description: "Background checks and monitoring configuration.",
    default: {},
});
