// @overvis/opcb-ts-shared/dist/schema/static-ip.schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.STATIC_IP_SCHEMA = void 0;
var typebox_1 = require("@sinclair/typebox");
exports.STATIC_IP_SCHEMA = typebox_1.Type.Object({
    ipAddress: typebox_1.Type.String({
        format: "ipv4",
        description: "IP address.",
        default: "192.168.0.111",
    }),
    subnetMask: typebox_1.Type.String({
        format: "ipv4",
        description: "Subnet mask.",
        default: "255.255.255.0",
    }),
    gateway: typebox_1.Type.String({
        format: "ipv4",
        description: "Gateway IP.",
        default: "192.168.0.1",
    }),
    primaryDns: typebox_1.Type.String({
        format: "ipv4",
        description: "Primary DNS.",
        default: "192.168.0.1",
    }),
    secondaryDns: typebox_1.Type.String({
        format: "ipv4",
        description: "Secondary DNS.",
        default: "8.8.8.8",
    }),
}, {
    description: "Static IP configuration.",
    additionalProperties: false,
    default: {},
});
