// @overvis/opcb-ts-shared/dist/redis/actor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestRedisClient = exports.TskRedisClient = exports.VirRedisClient = exports.ApiRedisClient = exports.LinRedisClient = exports.CfgRedisClient = exports.ActorRedisClient = void 0;
var tslib_1 = require("tslib");
var ActorRedisClient = /** @class */ (function () {
    function ActorRedisClient(client, actor) {
        var _this = this;
        this.client = client;
        this.actor = actor;
        this.getConfs = this.client.getConfs.bind(this.client);
        this.getState = this.client.getState.bind(this.client);
        this.setState = this.client.setState.bind(this.client);
        this.sendMsg = function (cmd) { return tslib_1.__awaiter(_this, void 0, void 0, function () { return tslib_1.__generator(this, function (_a) {
            return [2 /*return*/, this.client.sendMsg(this.actor, cmd)];
        }); }); };
        this.sendRequest = function (req, dst) { return tslib_1.__awaiter(_this, void 0, void 0, function () { return tslib_1.__generator(this, function (_a) {
            return [2 /*return*/, this.client.sendRequest(req, this.actor, dst)];
        }); }); };
        this.registerRequestHandler = function (cmd, handler) { return _this.client.registerRequestHandler(_this.actor, cmd, handler); };
    }
    return ActorRedisClient;
}());
exports.ActorRedisClient = ActorRedisClient;
var CfgRedisClient = /** @class */ (function (_super) {
    tslib_1.__extends(CfgRedisClient, _super);
    function CfgRedisClient() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.setConfs = _this.client.setConfs.bind(_this.client);
        return _this;
    }
    return CfgRedisClient;
}(ActorRedisClient));
exports.CfgRedisClient = CfgRedisClient;
var LinRedisClient = /** @class */ (function (_super) {
    tslib_1.__extends(LinRedisClient, _super);
    function LinRedisClient() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.setList = _this.client.setList.bind(_this.client);
        return _this;
    }
    return LinRedisClient;
}(ActorRedisClient));
exports.LinRedisClient = LinRedisClient;
var ApiRedisClient = /** @class */ (function (_super) {
    tslib_1.__extends(ApiRedisClient, _super);
    function ApiRedisClient() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.listenLogStream = _this.client.listenLogStream.bind(_this.client);
        _this.waitForDeviceStateUpdate = function () { return tslib_1.__awaiter(_this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.client.waitForMsg("dev-state")];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        }); };
        _this.getList = _this.client.getList.bind(_this.client);
        _this.createSession = _this.client.createSession.bind(_this.client);
        _this.deleteSession = _this.client.deleteSession.bind(_this.client);
        return _this;
    }
    return ApiRedisClient;
}(ActorRedisClient));
exports.ApiRedisClient = ApiRedisClient;
var VirRedisClient = /** @class */ (function (_super) {
    tslib_1.__extends(VirRedisClient, _super);
    function VirRedisClient() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.writeLogStream = _this.client.writeLogStream.bind(_this.client);
        _this.registerPartialRequestHandler = function (cmd, handler) { return _this.client.registerRequestHandler(_this.actor, cmd, handler); };
        return _this;
    }
    return VirRedisClient;
}(ActorRedisClient));
exports.VirRedisClient = VirRedisClient;
var TskRedisClient = /** @class */ (function (_super) {
    tslib_1.__extends(TskRedisClient, _super);
    function TskRedisClient() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.writeTaskLogRec = function (taskId, level, payload) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                return [2 /*return*/, this.client.writeGeneralLogStream("taskLog:" + taskId, level, payload)];
            });
        }); };
        return _this;
    }
    return TskRedisClient;
}(ActorRedisClient));
exports.TskRedisClient = TskRedisClient;
var TestRedisClient = /** @class */ (function (_super) {
    tslib_1.__extends(TestRedisClient, _super);
    function TestRedisClient() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.setConfs = _this.client.setConfs.bind(_this.client);
        _this.waitForNextMsg = _this.client.waitForNextMsg.bind(_this.client);
        _this.getLastMsgId = _this.client.getLastMsgId.bind(_this.client);
        _this.getStreamInfo = _this.client.getStreamInfo.bind(_this.client);
        return _this;
    }
    return TestRedisClient;
}(ActorRedisClient));
exports.TestRedisClient = TestRedisClient;
