// @overvis/opcb-ts-shared/dist/redis/msg.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializeObjToMsg = exports.getMsgFromFields = exports.CmdStreamMessage = void 0;
var CmdStreamMessage = /** @class */ (function () {
    function CmdStreamMessage() {
    }
    CmdStreamMessage.fromMsg = function (_msg) {
        throw new Error("unimplemented");
    };
    return CmdStreamMessage;
}());
exports.CmdStreamMessage = CmdStreamMessage;
function getMsgFromFields(fields) {
    var msg = {};
    for (var i = 0; i < fields.length; i += 2) {
        var key = fields[i].toString();
        msg[key] = fields[i + 1];
    }
    if (!("CMD" in msg)) {
        throw new Error("No CMD in command stream message: " + JSON.stringify(msg));
    }
    if (!("SRC" in msg)) {
        throw new Error("No SRC in command stream message: " + JSON.stringify(msg));
    }
    var res = {
        CMD: msg.CMD.toString(),
        SRC: msg.SRC.toString(),
        msg: msg,
    };
    if (msg.DST) {
        res.DST = msg.DST.toString();
    }
    if (msg.REQID) {
        res.REQID = msg.REQID.toString();
    }
    return res;
}
exports.getMsgFromFields = getMsgFromFields;
function serializeObjToMsg(obj, src, dst, reqId) {
    var msg = obj.toMsg();
    var res = {
        CMD: obj.cmd,
        SRC: src,
        msg: msg,
    };
    if (dst) {
        res.DST = dst;
    }
    if (reqId) {
        res.REQID = reqId.toString();
    }
    return res;
}
exports.serializeObjToMsg = serializeObjToMsg;
