// @overvis/opcb-api/dist/model/lora-sens/bind.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.bindDevice = void 0;
const sku_1 = require("./sku");
function bindDevice(queries, did, devKindCode, firmwareVersion, frequencyDeviceKhz, frequencyGatewayKhz, bandwidthKhz, spreadFactor, codingRateBits) {
    const sku = (0, sku_1.deductLoraSku)(devKindCode);
    queries.transaction(() => {
        // insert into table devices new device
        const deviceId = queries.device.insert({
            title: `LoRa Sensor ${sku}`,
            origTemplateId: sku,
            channel: `LGW:${did}`,
        });
        // insert into table lorasens_devices new lorasens device
        queries.loraSens.insertDevice({
            deviceId,
            did,
            sku,
            deviceKindCode: devKindCode,
            firmwareVersion,
            password: Math.floor(Math.random() * 100000000)
                .toString()
                .padStart(8, "0"),
            updatePeriodSec: 300,
            activeModeTill: Math.floor(new Date().getTime() / 1000),
            frequencyDeviceKhz,
            frequencyGatewayKhz,
            bandwidthKhz,
            spreadingFactor: spreadFactor,
            codingRateBits,
        });
    })();
}
exports.bindDevice = bindDevice;
