// @overvis/opcb-virtual-device/dist/modbus.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createModbusError = exports.ModbusFunctionCode = exports.ModbusExceptionCode = void 0;
var ModbusExceptionCode;
(function (ModbusExceptionCode) {
    ModbusExceptionCode[ModbusExceptionCode["IllegalFunction"] = 1] = "IllegalFunction";
    ModbusExceptionCode[ModbusExceptionCode["IllegalDataAddress"] = 2] = "IllegalDataAddress";
    ModbusExceptionCode[ModbusExceptionCode["IllegalDataValue"] = 3] = "IllegalDataValue";
    ModbusExceptionCode[ModbusExceptionCode["SlaveDeviceFailure"] = 4] = "SlaveDeviceFailure";
})(ModbusExceptionCode = exports.ModbusExceptionCode || (exports.ModbusExceptionCode = {}));
var ModbusFunctionCode;
(function (ModbusFunctionCode) {
    ModbusFunctionCode[ModbusFunctionCode["ReadHoldingRegisters"] = 3] = "ReadHoldingRegisters";
    ModbusFunctionCode[ModbusFunctionCode["ReadInputRegisters"] = 4] = "ReadInputRegisters";
    ModbusFunctionCode[ModbusFunctionCode["WriteSingleHolding"] = 6] = "WriteSingleHolding";
    ModbusFunctionCode[ModbusFunctionCode["WriteMultipleHolding"] = 16] = "WriteMultipleHolding";
})(ModbusFunctionCode = exports.ModbusFunctionCode || (exports.ModbusFunctionCode = {}));
function createModbusError(func, exceptionCode) {
    const response = Buffer.alloc(2);
    response.writeUInt8(func + 0x80, 0);
    response.writeUInt8(exceptionCode, 1);
    return response;
}
exports.createModbusError = createModbusError;
