// @overvis/opcb-api/dist/endpoints/list-wifi-aps.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handler = exports.SCHEMA = void 0;
const opcb_ts_shared_1 = require("@overvis/opcb-ts-shared");
const server_tools_1 = require("@overvis/server-tools");
const typebox_1 = require("@sinclair/typebox");
const auth_1 = require("../auth");
const OK_RESPONSE_SCHEMA = typebox_1.Type.Array(typebox_1.Type.Object({
    ssid: typebox_1.Type.String(),
    bssid: typebox_1.Type.String(),
    signal: typebox_1.Type.Number(),
    freq: typebox_1.Type.Number(),
}));
exports.SCHEMA = {
    response: (0, server_tools_1.apiResponse)(OK_RESPONSE_SCHEMA),
};
async function openHandler(_req) {
    await this.bus.sendRequest(new opcb_ts_shared_1.redisCmd.UpdateList("aps"));
    const apStrs = await this.bus.getList("aps");
    return apStrs.map((s) => {
        const [ssid, bssid, signal, freq] = s.split("|");
        return { ssid, bssid, signal: Number(signal), freq: Number(freq) };
    });
}
exports.handler = (0, auth_1.auth)("readConfs", openHandler);
