// @overvis/opcb-api/dist/endpoints/lora-sens-get-weight-modules-data.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handler = exports.SCHEMA = void 0;
const server_tools_1 = require("@overvis/server-tools");
const typebox_1 = require("@sinclair/typebox");
const device_1 = require("../model/lora-sens/device");
const weight_sensor_1 = require("../model/lora-sens/weight-sensor");
const QUERYSTRING_SCHEMA = typebox_1.Type.Object({
    did: typebox_1.Type.Number(),
    deviceId: typebox_1.Type.String(),
});
const OK_RESPONSE_SCHEMA = typebox_1.Type.Array(typebox_1.Type.Object({
    moduleNumber: typebox_1.Type.Number(),
    weight: typebox_1.Type.Optional(typebox_1.Type.Number()),
    moduleError: typebox_1.Type.Optional(typebox_1.Type.String()),
    state: typebox_1.Type.Union([typebox_1.Type.Literal("on"), typebox_1.Type.Literal("off"), typebox_1.Type.Literal("processing")]),
    deviceId: typebox_1.Type.String(),
    did: typebox_1.Type.Number(),
}));
exports.SCHEMA = {
    querystring: QUERYSTRING_SCHEMA,
    response: (0, server_tools_1.apiResponse)(OK_RESPONSE_SCHEMA),
};
function handler(req) {
    (0, device_1.setActiveDevice)(this.queries, req.query.did);
    const resp = (0, weight_sensor_1.getLoraSensWeightModules)(this.queries, req.query.did, req.query.deviceId);
    return resp;
}
exports.handler = handler;
