// @overvis/opcb-api/dist/endpoints/lora-sens-get-temperature-settings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handler = exports.SCHEMA = void 0;
const server_tools_1 = require("@overvis/server-tools");
const typebox_1 = require("@sinclair/typebox");
const device_1 = require("../model/lora-sens/device");
const temperature_sensor_1 = require("../model/lora-sens/temperature-sensor");
const QUERYSTRING_SCHEMA = typebox_1.Type.Object({
    deviceId: typebox_1.Type.String(),
    did: typebox_1.Type.Number(),
});
const OK_RESPONSE_SCHEMA = typebox_1.Type.Object({
    temperatureOffset: typebox_1.Type.Object({
        value: typebox_1.Type.Optional(typebox_1.Type.Number()),
        lastUpdatedOnSec: typebox_1.Type.Optional(typebox_1.Type.Number()),
    }),
    humidityOffset: typebox_1.Type.Object({
        value: typebox_1.Type.Optional(typebox_1.Type.Number()),
        lastUpdatedOnSec: typebox_1.Type.Optional(typebox_1.Type.Number()),
    }),
    title: typebox_1.Type.String(),
    updatePeriodSec: typebox_1.Type.Integer(),
});
exports.SCHEMA = {
    querystring: QUERYSTRING_SCHEMA,
    response: (0, server_tools_1.apiResponse)(OK_RESPONSE_SCHEMA),
};
function handler(req) {
    (0, device_1.setActiveDevice)(this.queries, req.query.did);
    const resp = (0, temperature_sensor_1.getLoraSensTemperatureSettings)(this.queries, req.query.did, req.query.deviceId);
    return resp;
}
exports.handler = handler;
