// @overvis/opcb-api/dist/endpoints/lora-sens-set-temperature-settings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handler = exports.SCHEMA = void 0;
const server_tools_1 = require("@overvis/server-tools");
const typebox_1 = require("@sinclair/typebox");
const opcb_ts_shared_1 = require("@overvis/opcb-ts-shared");
const temperature_sensor_1 = require("../model/lora-sens/temperature-sensor");
const BODY_SCHEMA = typebox_1.Type.Object({
    deviceId: typebox_1.Type.String(),
    did: typebox_1.Type.Number(),
    temperatureOffset: typebox_1.Type.Optional(typebox_1.Type.Number({ minimum: -10, maximum: 10 })),
    humidityOffset: typebox_1.Type.Optional(typebox_1.Type.Number({
        minimum: -10,
        maximum: 10,
    })),
    title: typebox_1.Type.Optional(typebox_1.Type.String({ minLength: 1, maxLength: 200 })),
    updatePeriodSec: typebox_1.Type.Optional(typebox_1.Type.Integer({
        minimum: 1,
        maximum: 36686400,
    })),
});
const OK_RESPONSE_SCHEMA = typebox_1.Type.Object({}, { additionalProperties: false });
exports.SCHEMA = {
    body: BODY_SCHEMA,
    response: (0, server_tools_1.apiResponse)(OK_RESPONSE_SCHEMA),
};
async function handler(req) {
    (0, temperature_sensor_1.setLoraSensTeperatureSettings)(this.queries, req.body.did, req.body.deviceId, req.body.temperatureOffset, req.body.humidityOffset, req.body.title, req.body.updatePeriodSec);
    await this.bus.sendMsg(new opcb_ts_shared_1.redisCmd.SyncDevices());
    return {};
}
exports.handler = handler;
