// @overvis/opcb-api/dist/endpoints/lora-sens-get-param-values.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handler = exports.SCHEMA = void 0;
const server_tools_1 = require("@overvis/server-tools");
const typebox_1 = require("@sinclair/typebox");
const device_1 = require("../model/lora-sens/device");
const param_1 = require("../model/lora-sens/param");
const sku_1 = require("../model/lora-sens/sku");
const QUERYSTRING_SCHEMA = typebox_1.Type.Object({
    deviceId: typebox_1.Type.String(),
    did: typebox_1.Type.Number(),
    sku: typebox_1.Type.Enum(sku_1.LoraSensSku),
});
const OK_RESPONSE_SCHEMA = typebox_1.Type.Array(typebox_1.Type.Object({
    id: typebox_1.Type.String(),
    name: typebox_1.Type.String(),
    value: typebox_1.Type.Optional(typebox_1.Type.String()),
    error: typebox_1.Type.Optional(typebox_1.Type.String()),
}));
exports.SCHEMA = {
    querystring: QUERYSTRING_SCHEMA,
    response: (0, server_tools_1.apiResponse)(OK_RESPONSE_SCHEMA),
};
function handler(req) {
    (0, device_1.setActiveDevice)(this.queries, req.query.did);
    const resp = (0, param_1.getLoraSensParamLastResults)(this.queries, req.query.deviceId, req.query.sku);
    return resp;
}
exports.handler = handler;
