// @overvis/opcb-api/dist/endpoints/lora-sens-get-readings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handler = exports.openHandler = exports.SCHEMA = void 0;
const server_tools_1 = require("@overvis/server-tools");
const typebox_1 = require("@sinclair/typebox");
const device_1 = require("../model/lora-sens/device");
const readings_1 = require("../model/lora-sens/readings");
const auth_1 = require("../auth");
const QUERYSTRING_SCHEMA = typebox_1.Type.Object({
    deviceId: typebox_1.Type.String(),
    params: typebox_1.Type.String(),
});
const OK_RESPONSE_SCHEMA = typebox_1.Type.Array(typebox_1.Type.Object({
    id: typebox_1.Type.String(),
    readingName: typebox_1.Type.String(),
    displayNameEn: typebox_1.Type.String(),
    value: typebox_1.Type.Optional(typebox_1.Type.String()),
    error: typebox_1.Type.Optional(typebox_1.Type.String()),
    unit: typebox_1.Type.String(),
}));
exports.SCHEMA = {
    querystring: QUERYSTRING_SCHEMA,
    response: (0, server_tools_1.apiResponse)(OK_RESPONSE_SCHEMA),
};
function openHandler(req) {
    const resp = (0, readings_1.getLastReadings)(this.queries, req.query.deviceId, req.query.params.split(","));
    (0, device_1.setDeviceActive)(this.queries, req.query.deviceId);
    return resp;
}
exports.openHandler = openHandler;
exports.handler = (0, auth_1.auth)("any", openHandler);
