// @overvis/opcb-api/dist/endpoints/lora-sens-get-weight-modules.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handler = exports.openHandler = exports.SCHEMA = void 0;
const server_tools_1 = require("@overvis/server-tools");
const typebox_1 = require("@sinclair/typebox");
const device_1 = require("../model/lora-sens/device");
const weight_module_1 = require("../model/lora-sens/weight-module");
const auth_1 = require("../auth");
const QUERYSTRING_SCHEMA = typebox_1.Type.Object({
    deviceId: typebox_1.Type.String(),
});
const OK_RESPONSE_SCHEMA = typebox_1.Type.Array(typebox_1.Type.Object({
    moduleNumber: typebox_1.Type.Number(),
    weight: typebox_1.Type.Optional(typebox_1.Type.Number()),
    moduleError: typebox_1.Type.Optional(typebox_1.Type.String()),
    state: typebox_1.Type.Union([typebox_1.Type.Literal("on"), typebox_1.Type.Literal("off"), typebox_1.Type.Literal("processing")]),
    deviceId: typebox_1.Type.String(),
    did: typebox_1.Type.Number(),
}));
exports.SCHEMA = {
    querystring: QUERYSTRING_SCHEMA,
    response: (0, server_tools_1.apiResponse)(OK_RESPONSE_SCHEMA),
};
function openHandler(req) {
    const resp = (0, weight_module_1.getWeightModules)(this.queries, req.query.deviceId);
    (0, device_1.setDeviceActive)(this.queries, req.query.deviceId);
    return resp;
}
exports.openHandler = openHandler;
exports.handler = (0, auth_1.auth)("readConfs", openHandler);
