// @overvis/opcb-api/dist/endpoints/lora-sens-set-weight-module-state.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handler = exports.openHandler = exports.SCHEMA = void 0;
const server_tools_1 = require("@overvis/server-tools");
const typebox_1 = require("@sinclair/typebox");
const weight_module_1 = require("../model/lora-sens/weight-module");
const auth_1 = require("../auth");
const BODY_SCHEMA = typebox_1.Type.Object({
    did: typebox_1.Type.Number(),
    moduleNumber: typebox_1.Type.Number({
        minimum: 0,
        maximum: 7,
    }),
    isEnabled: typebox_1.Type.Boolean(),
});
const OK_RESPONSE_SCHEMA = typebox_1.Type.Object({}, { additionalProperties: false });
exports.SCHEMA = {
    body: BODY_SCHEMA,
    response: (0, server_tools_1.apiResponse)(OK_RESPONSE_SCHEMA),
};
function openHandler(req) {
    (0, weight_module_1.setWeightModuleState)(this.queries, req.body.did, req.body.moduleNumber, req.body.isEnabled);
    return {};
}
exports.openHandler = openHandler;
exports.handler = (0, auth_1.auth)("writeConfs", openHandler);
