// @overvis/opcb-api/dist/endpoints/lora-sens-device-save-settings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handler = exports.SCHEMA = void 0;
const server_tools_1 = require("@overvis/server-tools");
const typebox_1 = require("@sinclair/typebox");
const opcb_ts_shared_1 = require("@overvis/opcb-ts-shared");
const device_1 = require("../model/lora-sens/device");
const auth_1 = require("../auth");
const BODY_SCHEMA = typebox_1.Type.Object({
    deviceId: typebox_1.Type.String(),
    name: typebox_1.Type.Optional(typebox_1.Type.String()),
    updatePeriodSec: typebox_1.Type.Optional(typebox_1.Type.Number()),
});
const OK_RESPONSE_SCHEMA = typebox_1.Type.Object({}, { additionalProperties: false });
exports.SCHEMA = {
    body: BODY_SCHEMA,
    response: (0, server_tools_1.apiResponse)(OK_RESPONSE_SCHEMA),
};
async function openHandler(req) {
    (0, device_1.saveDeviceSettings)(req.body.deviceId, this.queries, req.body.name, req.body.updatePeriodSec);
    await this.bus.sendMsg(new opcb_ts_shared_1.redisCmd.SyncDevices());
    return {};
}
exports.handler = (0, auth_1.auth)("readConfs", openHandler);
