// @overvis/opcb-api/dist/endpoints/mbtcp-clients.js
"use strict";
/*
TOKEN=`curl -s -H "Content-Type: application/json" \
        -d '{"username": "admin", "password": "admin" }' \
        "http://localhost:3000/login/" | \
    jq -r ".token"` \
&& \
curl -s -S -H "Content-Type: application/json" -H "Authorization: bearer $TOKEN" \
    "http://localhost:3000/mbtcp-clients/" | \
    jq -C
*/
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.REMOVE_SCHEMA = exports.UPDATE_SCHEMA = exports.CREATE_SCHEMA = exports.LIST_SCHEMA = exports.remove = exports.update = exports.create = exports.list = void 0;
const tslib_1 = require("tslib");
/*
TOKEN=`curl -s -H "Content-Type: application/json" \
        -d '{"username": "admin", "password": "admin" }' \
        "http://localhost:3000/login/" | \
    jq -r ".token"` \
&& \
curl -s -S -H "Content-Type: application/json" -H "Authorization: bearer $TOKEN" \
    -d '{}' \
    "http://localhost:3000/mbtcp-client/127.0.0.1-502/remove/" | \
    jq -C
*/
const typebox_1 = require("@sinclair/typebox");
const conf_crud_endpoint_1 = require("../conf-crud-endpoint");
const mbtcpClients = tslib_1.__importStar(require("../model/mptcp-clients"));
_a = (0, conf_crud_endpoint_1.createConfCrudEndpoints)(mbtcpClients, typebox_1.Type.Object({
    ip: typebox_1.Type.String(),
    port: typebox_1.Type.String(),
}), (p) => ({
    "remoteServer:ip": p.ip,
    "remoteServer:port": p.port,
}), mbtcpClients.LIST_ITEM_SCHEMA, mbtcpClients.CREATE_ITEM_SCHEMA, mbtcpClients.UPDATE_ITEM_SCHEMA), exports.list = _a.list, exports.create = _a.create, exports.update = _a.update, exports.remove = _a.remove, exports.LIST_SCHEMA = _a.LIST_SCHEMA, exports.CREATE_SCHEMA = _a.CREATE_SCHEMA, exports.UPDATE_SCHEMA = _a.UPDATE_SCHEMA, exports.REMOVE_SCHEMA = _a.REMOVE_SCHEMA;
