// @overvis/opcb-api/dist/endpoints/mf-restore.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handler = exports.getVpnDeviceInfo = exports.getDeviceIpFromMotherland = exports.SCHEMA = void 0;
const opcb_ts_shared_1 = require("@overvis/opcb-ts-shared");
const server_tools_1 = require("@overvis/server-tools");
const typebox_1 = require("@sinclair/typebox");
const __1 = require("..");
const auth_1 = require("../auth");
const motherland_1 = require("../model/motherland");
const mf_register_1 = require("./mf-register");
const BODY_SCHEMA = typebox_1.Type.Object({
    privateKey: typebox_1.Type.String(),
});
const OK_RESPONSE_SCHEMA = typebox_1.Type.Object({}, { additionalProperties: false });
exports.SCHEMA = {
    body: BODY_SCHEMA,
    response: (0, server_tools_1.apiResponse)(OK_RESPONSE_SCHEMA),
};
/*
curl -d '{"privateKey": "IHIP51hQdamhHj4hoasc5NLWDlfjdPfk0mtgllg8iFo="}'  -H "Content-Type: application/json" http://localhost/api/manufacture/restore/
*/
async function openHandler(req) {
    // get public key
    const privateKey = req.body.privateKey;
    const resp = await this.bus.sendRequest(new opcb_ts_shared_1.redisCmd.GenPubKey(privateKey));
    if ("error" in resp.result) {
        throw new server_tools_1.ApiUserError(400, "IncorrectPrivateKey", "Failed to generate public key.");
    }
    const publicKey = resp.result.publicKey;
    const defaultUserName = "admin";
    const { "overvisVpn:publicMotherlandApiUrl": motherlandApiUrl } = await this.bus.getConfs([
        "overvisVpn:publicMotherlandApiUrl",
    ]);
    if (!motherlandApiUrl) {
        throw new Error("Failed to get motherland api url.");
    }
    // getting ip and setting up vpn
    const ipResult = await getDeviceIpFromMotherland(motherlandApiUrl, publicKey);
    const allowedIps = ipResult.motherlandVpnIpAddress.split(".").slice(0, 2).join(".") + ".0.0/16";
    let confRes = await this.bus.sendRequest(new opcb_ts_shared_1.redisCmd.SetConfig({
        "overvisVpn:wireguard:privateKey": privateKey,
        "overvisVpn:wireguard:endpointPublicKey": ipResult.motherlandWgPublicKey,
        "overvisVpn:wireguard:vpnMotherlandServerIp": ipResult.motherlandVpnIpAddress,
        "overvisVpn:wireguard:ipAddressCidr": ipResult.vpnIpAddress + "/16",
        "overvisVpn:wireguard:endpointAllowedIpsCidr": allowedIps,
    }, false, "both"));
    if (confRes.status !== "ok") {
        throw new Error("Failed to set config.");
    }
    await (0, server_tools_1.timeout)(20000, this.bus.waitForDeviceStateUpdate());
    // getting registration info
    const infoResult = await getVpnDeviceInfo(ipResult.motherlandVpnApiUrl);
    if (!infoResult.defaultPassword || infoResult.defaultPassword.length < 8) {
        throw new Error(`Motherland returned incorrect default password: ${infoResult.defaultPassword || "null"}`);
    }
    // update config and factory config from motherland api response
    confRes = await this.bus.sendRequest(new opcb_ts_shared_1.redisCmd.SetConfig({
        "overvisVpn:wireguard:privateKey": privateKey,
        "overvisVpn:wireguard:endpointPublicKey": infoResult.motherlandWgPublicKey,
        "overvisVpn:wireguard:vpnMotherlandServerIp": infoResult.motherlandVpnIpAddress,
        "overvisVpn:wireguard:ipAddressCidr": infoResult.vpnIpAddress + "/16",
        "overvisVpn:wireguard:endpointAllowedIpsCidr": allowedIps,
        "overvisVpn:addRootSshKey": infoResult.motherlandSshPublicKey,
        "overvisVpn:pinCode": infoResult.pinCode,
        "global:model": infoResult.sku,
        "global:manufacturerName": infoResult.manufacturerName,
        "global:modelName": infoResult.labelName,
        "global:labelLink": infoResult.labelLink,
        "elan:mac": infoResult.macAddress,
    }, false, "both"));
    if (confRes.status !== "ok") {
        throw new Error("Failed to set config.");
    }
    confRes = await this.bus.sendRequest(new opcb_ts_shared_1.redisCmd.SetConfig({
        "auth:users:0:username": defaultUserName,
        "auth:users:0:password": null,
        "auth:users:0:encryptedPassword": (0, opcb_ts_shared_1.encryptPassword)(infoResult.defaultPassword),
        "auth:users:0:modbusDevicesAccess": "readWrite",
        "auth:users:0:configurationAccess": "readWrite",
        "auth:users:0:allowAuthThroughApi": "true",
        "auth:users:0:allowAuthThroughModbus": "false",
        "wlan:apSettings:password": (0, opcb_ts_shared_1.encryptWifiPassword)(infoResult.defaultPassword),
    }, false, "factory"));
    if (confRes.status !== "ok") {
        throw new Error("Failed to set factory config.");
    }
    // request re-rendering of the label image
    await this.bus.sendRequest(new opcb_ts_shared_1.redisCmd.RerenderLabelImage(defaultUserName, infoResult.defaultPassword));
    return {};
}
const MOTHERLAND_DEVICE_IP_SCHEMA = typebox_1.Type.Object({
    motherlandWgPublicKey: typebox_1.Type.String(),
    motherlandVpnIpAddress: typebox_1.Type.String(),
    motherlandVpnApiUrl: typebox_1.Type.String(),
    vpnIpAddress: typebox_1.Type.String(),
});
async function getDeviceIpFromMotherland(motherlandApiUrl, deviceWgPublicKey) {
    __1.logger.info("Getting device VPN IP from Motherland...");
    return await (0, motherland_1.motherlandRequest)(`${motherlandApiUrl}/vpn-device-ip/`, { method: "GET", params: { deviceWgPublicKey } }, MOTHERLAND_DEVICE_IP_SCHEMA);
}
exports.getDeviceIpFromMotherland = getDeviceIpFromMotherland;
async function getVpnDeviceInfo(motherlandVpnApiUrl) {
    return await (0, motherland_1.motherlandRequest)(`${motherlandVpnApiUrl}/vpn-device-info/`, { method: "GET" }, mf_register_1.MOTHERLAND_REGISTRATION_INFO_SCHEMA);
}
exports.getVpnDeviceInfo = getVpnDeviceInfo;
exports.handler = (0, auth_1.auth)("writeConfs", openHandler);
