// @overvis/opcb-api/dist/endpoints/modbus-free-ranges.js
"use strict";
/*
TOKEN=`curl -s -H "Content-Type: application/json" \
        -d '{"username": "admin", "password": "admin" }' \
        "http://localhost:3000/login/" | \
    jq -r ".token"` \
&& \
curl -s -S -H "Content-Type: application/json" -H "Authorization: bearer $TOKEN" \
    "http://localhost:3000/modbus/free-ranges/" | \
    jq -C
*/
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.REMOVE_SCHEMA = exports.UPDATE_SCHEMA = exports.CREATE_SCHEMA = exports.LIST_SCHEMA = exports.remove = exports.update = exports.create = exports.list = void 0;
const tslib_1 = require("tslib");
/*
TOKEN=`curl -s -H "Content-Type: application/json" \
        -d '{"username": "admin", "password": "admin" }' \
        "http://localhost:3000/login/" | \
    jq -r ".token"` \
&& \
curl -s -S -H "Content-Type: application/json" -H "Authorization: bearer $TOKEN" \
    -d '{ "device": 22, "function": 6, "range": "100-200" }' \
    "http://localhost:3000/modbus/free-ranges/create/" | \
    jq -C

TOKEN=`curl -s -H "Content-Type: application/json" \
        -d '{"username": "admin", "password": "admin" }' \
        "http://localhost:3000/login/" | \
    jq -r ".token"` \
&& \
curl -s -S -H "Content-Type: application/json" -H "Authorization: bearer $TOKEN" \
    -d '{ "device": null, "function": 6, "range": "100-200,300-400" }' \
    "http://localhost:3000/modbus/free-ranges/create/" | \
    jq -C
*/
/*
TOKEN=`curl -s -H "Content-Type: application/json" \
        -d '{"username": "admin", "password": "admin" }' \
        "http://localhost:3000/login/" | \
    jq -r ".token"` \
&& \
curl -s -S -H "Content-Type: application/json" -H "Authorization: bearer $TOKEN" \
    -d '{}' \
    "http://localhost:3000/modbus/free-range/22-6/100-200/remove/" | \
    jq -C

TOKEN=`curl -s -H "Content-Type: application/json" \
        -d '{"username": "admin", "password": "admin" }' \
        "http://localhost:3000/login/" | \
    jq -r ".token"` \
&& \
curl -s -S -H "Content-Type: application/json" -H "Authorization: bearer $TOKEN" \
    -d '{}' \
    "http://localhost:3000/modbus/free-range/null-6/100-200,300-400/remove/" | \
    jq -C
*/
const typebox_1 = require("@sinclair/typebox");
const conf_crud_endpoint_1 = require("../conf-crud-endpoint");
const modbusFreeRanges = tslib_1.__importStar(require("../model/modbus-free-ranges"));
_a = (0, conf_crud_endpoint_1.createConfCrudEndpoints)(modbusFreeRanges, typebox_1.Type.Object({
    deviceFunction: typebox_1.Type.String(),
    range: typebox_1.Type.String(),
}), (p) => {
    const d = p.deviceFunction.split("-")[0];
    return {
        device: d === "null" ? undefined : d,
        function: p.deviceFunction.split("-")[1],
        registersRange: p.range,
    };
}, modbusFreeRanges.ITEM_SCHEMA, modbusFreeRanges.ITEM_SCHEMA, modbusFreeRanges.ITEM_SCHEMA), exports.list = _a.list, exports.create = _a.create, exports.update = _a.update, exports.remove = _a.remove, exports.LIST_SCHEMA = _a.LIST_SCHEMA, exports.CREATE_SCHEMA = _a.CREATE_SCHEMA, exports.UPDATE_SCHEMA = _a.UPDATE_SCHEMA, exports.REMOVE_SCHEMA = _a.REMOVE_SCHEMA;
