// @overvis/opcb-api/dist/endpoints/rs485.js
"use strict";
/*
TOKEN=`curl -s -H "Content-Type: application/json" \
        -d '{"username": "admin", "password": "admin" }' \
        "http://localhost:3000/login/" | \
    jq -r ".token"` \
&& \
curl -s -S -H "Content-Type: application/json" -H "Authorization: bearer $TOKEN" \
    "http://localhost:3000/rs485-connections/" | \
    jq -C
*/
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.REMOVE_SCHEMA = exports.UPDATE_SCHEMA = exports.CREATE_SCHEMA = exports.LIST_SCHEMA = exports.remove = exports.update = exports.create = exports.list = void 0;
const tslib_1 = require("tslib");
const typebox_1 = require("@sinclair/typebox");
const conf_crud_endpoint_1 = require("../conf-crud-endpoint");
const rs485 = tslib_1.__importStar(require("../model/rs485"));
_a = (0, conf_crud_endpoint_1.createConfCrudEndpoints)(rs485, typebox_1.Type.Object({
    deviceName: typebox_1.Type.String(),
}), (p) => p.deviceName, rs485.LIST_ITEM_SCHEMA, rs485.CREATE_ITEM_SCHEMA, rs485.UPDATE_ITEM_SCHEMA), exports.list = _a.list, exports.create = _a.create, exports.update = _a.update, exports.remove = _a.remove, exports.LIST_SCHEMA = _a.LIST_SCHEMA, exports.CREATE_SCHEMA = _a.CREATE_SCHEMA, exports.UPDATE_SCHEMA = _a.UPDATE_SCHEMA, exports.REMOVE_SCHEMA = _a.REMOVE_SCHEMA;
