// @overvis/opcb-api/dist/endpoints/state-get.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handler = exports.SCHEMA = void 0;
const opcb_ts_shared_1 = require("@overvis/opcb-ts-shared");
const server_tools_1 = require("@overvis/server-tools");
const typebox_1 = require("@sinclair/typebox");
const auth_1 = require("../auth");
const QUERYSTRING_SCHEMA = typebox_1.Type.Object({
    keys: typebox_1.Type.Optional(typebox_1.Type.String()),
}, { additionalProperties: false });
const OK_RESPONSE_SCHEMA = typebox_1.Type.Record(typebox_1.Type.String(), typebox_1.Type.String());
exports.SCHEMA = {
    querystring: QUERYSTRING_SCHEMA,
    response: (0, server_tools_1.apiResponse)(OK_RESPONSE_SCHEMA),
};
async function openHandler(req) {
    await this.bus.sendMsg(new opcb_ts_shared_1.redisCmd.EnableUserActiveMode(10000));
    let keys = [];
    if (req.query.keys) {
        keys = req.query.keys.split(",");
    }
    const res = await this.bus.getState(keys.length > 0 ? keys : undefined);
    return Object.fromEntries(Object.entries(res).filter((v) => v[1] !== undefined));
}
exports.handler = (0, auth_1.auth)("readConfs", openHandler);
