// @overvis/opcb-api/dist/endpoints/tasks-init.js
"use strict";
/*
TOKEN=`curl -s -H "Content-Type: application/json" \
        -d '{"username": "admin", "password": "admin" }' \
        "http://localhost:4004/api/login/" | \
    jq -r ".token"` \
&& \
curl -s -S -d "{}" -H "Content-Type: application/json" \
    -H "Authorization: bearer $TOKEN" \
    "http://localhost:4004/api/tasks/recompile/" | \
    jq -C
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.handler = exports.SCHEMA = void 0;
const opcb_ts_shared_1 = require("@overvis/opcb-ts-shared");
const server_tools_1 = require("@overvis/server-tools");
const typebox_1 = require("@sinclair/typebox");
const auth_1 = require("../auth");
const BODY_SCHEMA = typebox_1.Type.Object({
    mode: typebox_1.Type.Union([
        typebox_1.Type.Literal("init-only"),
        typebox_1.Type.Literal("update"),
        typebox_1.Type.Literal("full-overwrite"),
        typebox_1.Type.Literal("npm-install"),
        typebox_1.Type.Literal("npm-rebuild"),
    ]),
});
const OK_RESPONSE_SCHEMA = typebox_1.Type.Union([
    typebox_1.Type.Object({ result: typebox_1.Type.Literal("ok") }),
    typebox_1.Type.Object({ result: typebox_1.Type.Literal("error"), error: typebox_1.Type.String() }),
]);
exports.SCHEMA = {
    body: BODY_SCHEMA,
    response: (0, server_tools_1.apiResponse)(OK_RESPONSE_SCHEMA),
};
async function openHandler(req) {
    const res = await this.bus.sendRequest(new opcb_ts_shared_1.redisCmd.InitTasksProject(req.body.mode));
    if (res.error) {
        return { result: "error", error: res.error };
    }
    return { result: "ok" };
}
exports.handler = (0, auth_1.auth)("writeConfs", openHandler);
