// @overvis/opcb-api/dist/endpoints/task-update.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handler = exports.SCHEMA = void 0;
const opcb_ts_shared_1 = require("@overvis/opcb-ts-shared");
const server_tools_1 = require("@overvis/server-tools");
const typebox_1 = require("@sinclair/typebox");
const auth_1 = require("../auth");
const task_get_runs_1 = require("./task-get-runs");
const BODY_SCHEMA = typebox_1.Type.Object({
    name: typebox_1.Type.String({ maxLength: 200 }),
    description: typebox_1.Type.String(),
    autorun: typebox_1.Type.Boolean(),
    autorunPeriodSec: typebox_1.Type.Integer({ minimum: 0 }),
    restartOnError: typebox_1.Type.Boolean(),
    logLevel: typebox_1.Type.Union([
        typebox_1.Type.Literal("silent"),
        typebox_1.Type.Literal("error"),
        typebox_1.Type.Literal("warn"),
        typebox_1.Type.Literal("info"),
        typebox_1.Type.Literal("debug"),
        typebox_1.Type.Literal("trace"),
    ]),
    reportLevel: typebox_1.Type.Union([typebox_1.Type.Literal("stat-only"), typebox_1.Type.Literal("full-log")]),
});
const OK_RESPONSE_SCHEMA = typebox_1.Type.Object({});
exports.SCHEMA = {
    body: BODY_SCHEMA,
    params: task_get_runs_1.PARAMS_SCHEMA,
    response: (0, server_tools_1.apiResponse)(OK_RESPONSE_SCHEMA),
};
async function openHandler(req) {
    const _res = this.db
        .prepare("update tasks set task_name=@name, task_description=@description, autorun=@autorun, autorun_period_sec=@autorunPeriodSec, restart_on_error=@restartOnError, log_level=@logLevel, report_level=@reportLevel where id = @taskId;")
        .run({
        name: req.body.name,
        description: req.body.description,
        autorun: req.body.autorun,
        autorunPeriodSec: req.body.autorunPeriodSec,
        restartOnError: req.body.restartOnError,
        logLevel: req.body.logLevel,
        reportLevel: req.body.reportLevel,
        taskId: req.params.id,
    });
    await this.bus.sendMsg(new opcb_ts_shared_1.redisCmd.SyncTasks());
    return {};
}
exports.handler = (0, auth_1.auth)("writeConfs", openHandler);
