// @overvis/opcb-api/dist/endpoints/tests-check.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handler = exports.SCHEMA = void 0;
const tslib_1 = require("tslib");
const server_tools_1 = require("@overvis/server-tools");
const typebox_1 = require("@sinclair/typebox");
const axios_1 = tslib_1.__importDefault(require("axios"));
const auth_1 = require("../auth");
const QUERYSTRING_SCHEMA = typebox_1.Type.Object({
    testRunId: typebox_1.Type.String({ format: "uuid" }),
}, { additionalProperties: false });
const OK_RESPONSE_SCHEMA = typebox_1.Type.Object({
    testRunId: typebox_1.Type.String({ format: "uuid" }),
    startedOn: typebox_1.Type.String({ format: "date-time" }),
    finishedOn: typebox_1.Type.Optional(typebox_1.Type.String({ format: "date-time" })),
    wasAborted: typebox_1.Type.Boolean(),
    tests: typebox_1.Type.Array(typebox_1.Type.Object({
        name: typebox_1.Type.String(),
        description: typebox_1.Type.String(),
        result: typebox_1.Type.Union([
            typebox_1.Type.Object({ status: typebox_1.Type.Literal("waiting") }),
            typebox_1.Type.Object({
                status: typebox_1.Type.Literal("running"),
                startedOn: typebox_1.Type.String({ format: "date-time" }),
            }),
            typebox_1.Type.Object({
                status: typebox_1.Type.Literal("aborted"),
                startedOn: typebox_1.Type.String({ format: "date-time" }),
                finishedOn: typebox_1.Type.String({ format: "date-time" }),
            }),
            typebox_1.Type.Object({
                status: typebox_1.Type.Literal("passed"),
                startedOn: typebox_1.Type.String({ format: "date-time" }),
                finishedOn: typebox_1.Type.String({ format: "date-time" }),
            }),
            typebox_1.Type.Object({
                status: typebox_1.Type.Literal("failed"),
                error: typebox_1.Type.String(),
                startedOn: typebox_1.Type.String({ format: "date-time" }),
                finishedOn: typebox_1.Type.String({ format: "date-time" }),
            }),
        ]),
    }, { additionalProperties: false })),
}, { additionalProperties: false });
exports.SCHEMA = {
    response: (0, server_tools_1.apiResponse)(OK_RESPONSE_SCHEMA),
    querystring: QUERYSTRING_SCHEMA,
};
async function openHandler(req) {
    const motherlandVpnIp = (await this.bus.getConfs(["overvisVpn:wireguard:vpnMotherlandServerIp"]))["overvisVpn:wireguard:vpnMotherlandServerIp"];
    if (!motherlandVpnIp) {
        throw new Error("Motherland VPN IP is not set.");
    }
    const res = await axios_1.default.post(`http://${motherlandVpnIp}/api/abort-device-check/`, {
        headers: { "Content-Type": "application/json" },
        data: req.body,
    });
    return res.data;
}
exports.handler = (0, auth_1.auth)("writeConfs", openHandler);
