// @overvis/opcb-api/dist/endpoints/version.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handler = exports.SCHEMA = void 0;
const server_tools_1 = require("@overvis/server-tools");
const typebox_1 = require("@sinclair/typebox");
const OK_RESPONSE_SCHEMA = typebox_1.Type.Object({
    runtime: typebox_1.Type.String({ maxLength: 30 }),
    api: typebox_1.Type.String({ maxLength: 30 }),
    imminentUpgrade: typebox_1.Type.Optional(typebox_1.Type.Boolean()),
    lastFailedUpgradeRef: typebox_1.Type.Optional(typebox_1.Type.String()),
}, { additionalProperties: false });
exports.SCHEMA = {
    response: (0, server_tools_1.apiResponse)(OK_RESPONSE_SCHEMA),
};
async function handler() {
    var _a;
    const data = await this.bus.getState([
        "release:version",
        "release:updateState",
        "release:lastUpgradeError",
        "release:lastFailedUpgradeRef",
    ]);
    const runtime = (_a = data["release:version"]) !== null && _a !== void 0 ? _a : "unknown";
    const imminentUpgrade = data["release:updateState"] === "imminent-upgrade";
    const res = {
        runtime,
        api: process.env.npm_package_version || "0.0.0",
    };
    if (imminentUpgrade) {
        res.imminentUpgrade = true;
    }
    if (data["release:lastUpgradeError"]) {
        res.lastFailedUpgradeRef = data["release:lastFailedUpgradeRef"] || "unknown ref";
    }
    return res;
}
exports.handler = handler;
