// @overvis/server-tools/dist/promises.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PromiseMonitor = exports.timeout = void 0;
async function timeout(ms, promise) {
    return new Promise((resolve, reject) => {
        let timedOut = false;
        const timer = setTimeout(() => {
            reject(new Error("timeout"));
            timedOut = true;
        }, ms);
        promise
            .then((res) => {
            if (timedOut) {
                return;
            }
            clearTimeout(timer);
            resolve(res);
        })
            .catch((err) => {
            if (timedOut) {
                return;
            }
            clearTimeout(timer);
            reject(err);
        });
    });
}
exports.timeout = timeout;
class PromiseMonitor {
    constructor(prm) {
        this._status = "pending";
        this._isPending = true;
        this._isResolved = false;
        this._isRejected = false;
        prm.then((result) => {
            this._status = "resolved";
            this._isResolved = true;
            this._isPending = false;
            this._result = result;
        }, (err) => {
            this._status = "rejected";
            this._isPending = false;
            this._isRejected = true;
            this._error = err instanceof Error ? err : new Error("unknown error");
        });
    }
    get status() {
        return this._status;
    }
    get isPending() {
        return this._isPending;
    }
    get isResolved() {
        return this._isResolved;
    }
    get isRejected() {
        return this._isRejected;
    }
    get error() {
        return this._error;
    }
    get result() {
        return this._result;
    }
}
exports.PromiseMonitor = PromiseMonitor;
