// @overvis/opcb-config-manager/dist/tests/utils.js
"use strict";
/* eslint-disable no-console */
/* eslint-disable @typescript-eslint/no-empty-function */
Object.defineProperty(exports, "__esModule", { value: true });
exports.runConfman = void 0;
const tslib_1 = require("tslib");
const opcb_ts_shared_1 = require("@overvis/opcb-ts-shared");
const fs = tslib_1.__importStar(require("fs"));
const opcbConfigManager = tslib_1.__importStar(require(".."));
async function runConfman(configFileMixin, silent = false, factoryConfigFile = "{}", manufacturerFile = "") {
    let logger = {
        level: "silent",
        fatal: () => { },
        warn: () => { },
        debug: () => { },
        trace: () => { },
        info: () => { },
        error: () => { },
        silent: () => { },
    };
    if (!silent) {
        logger = {
            level: "warn",
            fatal: console.error,
            warn: console.warn,
            debug: () => { },
            trace: () => { },
            info: () => { },
            error: console.error,
            silent: () => { },
        };
    }
    const bus = new opcb_ts_shared_1.RedisBus("redis://127.0.0.1:6379/0", "tests", logger);
    await bus.clearCmdStream();
    const configFileStr = fs.readFileSync("./src/tests/test.config.json", "utf-8");
    const configStr = JSON.stringify(Object.assign(JSON.parse(configFileStr), configFileMixin), null, 4);
    const configFile = { contents: configStr };
    const [proc, stopConfman] = await opcbConfigManager.run(logger, {
        redisConnectString: "redis://127.0.0.1:6379/0",
        manufacturerFile: { contents: manufacturerFile },
        configFile,
        factoryConfigFile: { contents: factoryConfigFile },
        sqliteDbPath: ":memory:",
        sqliteLibDir: process.env.SQLITE_LIB_DIR || "./sqlite",
        dbMigrationsDir: "./db-migrations",
        testMode: true,
    });
    proc.catch(console.error);
    bus.listenCmdStreamForever().catch(console.error);
    return [
        bus,
        () => {
            stopConfman();
            bus.abortListen();
        },
        configFile,
    ];
}
exports.runConfman = runConfman;
